### PR TITLE
Full implementation of artifacts library

### PR DESCRIPTION
**Pull request by pasinskim**
`writer` implements writing part. It scans directories in provided location and looks for correct update directories (each needs to be of `sprintf("%04d")` format and there must be `meta-data`, `type-info` and `data/[update_name]` files).

Based on update type (specified in `type-info`) correct parser is chosen  to parse given type of update.
This design will allow extending library and add new parsers while there will be a need for doing so.
So far we have `rootfs-image` parser implemented, for parsing artifact file as described here: https://github.com/mendersoftware/meta-mender/blob/master/Documentation/artifact-format.md#files

All parser management is done by `parser` module. Specifically there is a `Parsers` structure managing registering and iterating over parsers while scanning updates.
Parsers MUST implement `ArtifactParser` interface which provides methods for writing and reading given type of update.

`reader` is the reader implementation. It provides reading whole artifact on the fly without storing any intermediate files. Also all operations are done using `io.Copy` calls which use `io.Reader` and 'io.Writer` interfaces underneath making sure that copying is performance and memory optimized. 

Similar as `writer`, `reader` uses parsers for parsing given update type.

`archiver` package provides convenient way of storing different data types into tar archive. So far we can store plain files, binary data and all JSON convertible data structures implemented in `metadata` package. 

`metadata` provides implementation of data structures shared between the `reader` and the `writer`. All the data structures in `metadata` are implementing `Validater` interface which in turn implements `io.Writer` interface. Together with `metadata_archiver` all the `metadata` data structures are implementing both `io.Reader` and `io.Writer` for convenient usage with `io.Copy`. 

Last, `reader` and `writer` are separated to limit the client binary size, as client needs only `reader` part.

### CODE REVIEW

File: archiver/stream_archiver.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+func (str *StreamArchiver) Archive(tw *tar.Writer) error {
+
+	hdr := &tar.Header{
+		Name: str.name,
+		Mode: 0600,
+		Size: int64(len(str.data)),
+	}
+	log.Debugf("arch: have header: %v", hdr)
```
Comment (maciejmrowiec) [inline archiver/stream_archiver.go]: I'm not sure if a stand alone library should be allowed to write logs - not something I would expect to happen.
This forces user to use logging and a specific one, I personally would be leaning towards leaving that to the user outside if and how he want to do this.

Comment (maciejmrowiec) [inline archiver/stream_archiver.go]: service uses context specific logger instances which would not work well with this

File: metadata/metadata.go [hunk 2]
```diff
@@ -47,6 +50,25 @@ func (i Info) Validate() error {
...
+		if err := dec.Decode(data); err != io.EOF {
+			break
+		} else if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (i *Info) Write(p []byte) (n int, err error) {
```
Comment (maciejmrowiec) [inline metadata/metadata.go]: Why using named returns if not using them?

File: metadata/metadata.go [hunk 3]
```diff
@@ -72,42 +94,55 @@ func (hi HeaderInfo) Validate() error {
...
 // Validate validates corectness of TypeInfo.
 func (ti TypeInfo) Validate() error {
-	if len(ti.Rootfs) == 0 {
+	if len(ti.Type) == 0 {
 		return ErrValidatingData
 	}
 	return nil
 }
 
+func (ti *TypeInfo) Write(p []byte) (n int, err error) {
```
Comment (maciejmrowiec) [inline metadata/metadata.go]: what is `p` ?

File: metadata/metadata.go [hunk 4]
```diff
@@ -72,42 +94,55 @@ func (hi HeaderInfo) Validate() error {
...
 	var deviceType interface{}
 	var imageID interface{}
 
-	for k, v := range m.data {
+	for k, v := range m {
 		if v == nil {
+			log.Error("empty element while validating metedata")
 			return ErrValidatingData
 		}
 		if strings.Compare(k, "DeviceType") == 0 {
```
Comment (maciejmrowiec) [inline metadata/metadata.go]: if this is a JSON key, probably would be more natural to lean towards lower case snake style

File: parser/parser_manager.go [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+package parser
+
+import (
+	"archive/tar"
+	"errors"
+	"io"
+)
+
+type ArtifactParser interface {
+	ParseHeader(tr *tar.Reader, hPath string) error
```
Comment (maciejmrowiec) [inline parser/parser_manager.go]: not sure if this is useful but I can see here reader/writer interfaces that would be composed into parser:

```
type ArticaftParseReader interface {
ParseHeader(tr *tar.Reader, hPath string) error
ParseData(r io.Reader) error
}

type ArticaftParseWriter interface {
ArchiveHeader(tw *tar.Writer, srcDir, dstDir string) error
ArchiveData(tw *tar.Writer, srcDir, dst string) error
}

type ArtifactParser interface {
ArticaftParseWriter
ArticaftParseReader
NeedsDataFile() bool
}
```

File: parser/rootfs_parser.go [hunk 6]
```diff
@@ -0,0 +1,346 @@
...
+	files    metadata.Files
+	tInfo    metadata.TypeInfo
+	metadata metadata.Metadata
+	updates  map[string]rootfsFile
+
+	sStore string
+	dStore io.Writer
+}
+
+func NewRootfsParser(sStoreDir string, w io.Writer) RootfsParser {
```
Comment (maciejmrowiec) [inline parser/rootfs_parser.go]: any particular reason why we pass a copy instead of reference ?

File: parser/rootfs_parser.go [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+	defer tw.Close()
+
+	for _, data := range rp.updates {
+		log.Infof("processing data file: %v [%v]", data.path, data.name)
+		a := archiver.NewFileArchiver(data.path, data.name)
+		if err := a.Archive(dtw); err != nil {
+			return err
+		}
+	}
+	//TODO
```
Comment (maciejmrowiec) [inline parser/rootfs_parser.go]: leftover todo

File: parser/rootfs_parser.go [hunk 8]
```diff
@@ -0,0 +1,346 @@
...
+		dStore:  w,
+		updates: map[string]rootfsFile{}}
+}
+
+func (rp RootfsParser) NeedsDataFile() bool {
+	return true
+}
+
+func (rp *RootfsParser) ArchiveData(tw *tar.Writer, srcDir, dst string) error {
+	f, err := os.Create("/tmp/my_data.tar.gz")
```
Comment (maciejmrowiec) [inline parser/rootfs_parser.go]: Why writing to tmpfs ? isn't this equivalent as keeping this in memory ?
What with about concurrent access if we have 2 instances of RootfsParser working on something?

For intermediate files this is useful: https://golang.org/pkg/io/ioutil/#TempFile

File: archiver/stream_archiver.go (hunk 1 above)
Comment (pasinskim) [inline archiver/stream_archiver.go]: Makes sense! I will remove logs.

File: metadata/metadata.go (hunk 2 above)
Comment (pasinskim) [inline metadata/metadata.go]: Looks like when you are trying to implement `io.Writer` interface go insist on using some special function signatures. Must be as is.

File: metadata/metadata.go (hunk 3 above)
Comment (pasinskim) [inline metadata/metadata.go]: See above. Must use a signature Go is proposing.

File: metadata/metadata.go (hunk 4 above)
Comment (pasinskim) [inline metadata/metadata.go]: I've been following this: https://github.com/mendersoftware/meta-mender/blob/master/Documentation/artifact-format.md#files

File: parser/parser_manager.go (hunk 5 above)
Comment (pasinskim) [inline parser/parser_manager.go]: Possibly, yes. But I want to make sure that you are providing both reader and writer to be used with this library.

File: parser/rootfs_parser.go (hunk 7 above)
Comment (pasinskim) [inline parser/rootfs_parser.go]: Some old merge crap.

File: parser/rootfs_parser.go (hunk 8 above)
Comment (pasinskim) [inline parser/rootfs_parser.go]: Like above. Merge issue.

File: parser/rootfs_parser.go (hunk 6 above)
Comment (pasinskim) [inline parser/rootfs_parser.go]: If you mean return, then no reason. I guess I can change to be more consistent.

**Comment by pasinskim** [comment]
@maciejmrowiec thanks for review. As this code is quite complicated (I think we might have over-engineered design of artifact file) I suggest leaving the logs for a while. 
Once we will integrate with other parts and we will see this working I will remove logs. For now those might be useful to fix some possible issues. And since those are here anyway, there is no difference if I will remove it now or in a week or so.

What do you think?

File: parser/rootfs_parser.go [hunk 9]
```diff
@@ -0,0 +1,346 @@
...
+
+	for _, data := range rp.updates {
+		log.Infof("processing data file: %v [%v]", data.path, data.name)
+		a := archiver.NewFileArchiver(data.path, data.name)
+		if err := a.Archive(dtw); err != nil {
+			return err
+		}
+	}
+	//TODO
+	dtw.Close()
```
Comment (maciejmrowiec) [inline parser/rootfs_parser.go]: it's already defered

File: parser/rootfs_parser.go [hunk 10]
```diff
@@ -0,0 +1,346 @@
...
+		return errors.Wrapf(err, "parser: can not store meta-data")
+	}
+
+	if err := archiveChecksums(tw, updFiles,
+		filepath.Join(srcDir, "data"),
+		filepath.Join(dstDir, "checksums")); err != nil {
+		return err
+	}
+
+	//TODO: get rid of bad Joins
```
Comment (maciejmrowiec) [inline parser/rootfs_parser.go]: leftover todo

File: parser/rootfs_parser.go [hunk 11]
```diff
@@ -0,0 +1,346 @@
...
+
+		relPath, err := filepath.Rel(hPath, hdr.Name)
+		if err != nil {
+			return err
+		}
+
+		log.Infof("processing rootfs image header file: %v %v", hdr.Name, hdr.Linkname)
+
+		switch {
+		case i == 0 && strings.Compare(relPath, "files") == 0:
```
Comment (maciejmrowiec) [inline parser/rootfs_parser.go]: may be a good idea to put fixed indexes in `const` (even a local one) to label what they mean

**Comment by maciejmrowiec** [comment]
Catch with logging is that once client or server use a version with those, it's going to vendor this, so removal of logs will go away with next dependency update whenever that may be.

**Comment by maciejmrowiec** [comment]
nice to have: godoc badge (so it's easy to find documentation)

**Comment by pasinskim** [comment]
@maciejmrowiec makes sense. I will remove logging then. 
What about the overall functionality? Do you think it is OK?

**Comment by maciejmrowiec** [comment]
it's slightly difficult to figure that out from diff :/
I know this may be a stretch but would be very helpful to get basic usage examples somewhere in doc (top section of the godoc is the part of the comments above `package xyz`) or readme which is quite common to have.

**Comment by pasinskim** [comment]
This is outdated.

**Pull request closed without merging**